rfcs/proposed/parallel_block_for_task_arena/README.md

```diff
@@ -0,0 +1,267 @@
+
+<img src="parallel_phase_state_final.png" width=800>
+
+Let's consider the semantics that an API for explicit parallel phases can provide:
+* Start of a parallel phase:
+  * Indicates the point from which the scheduler can use a hint and keep threads in the arena
+    for longer.
+  * Serves as a warm-up hint to the scheduler:
+    * Allows reducing computation start delays by initationg the wake-up of worker threads
+      in advance.
+* "Parallel phase" itself:
+  * Scheduler can implement different policies to retain threads in the arena.
+    * For instance, more aggressive policy might be implemented for _parallel phase_.
+      It can be beneficial in cases when the default arena leave policy is not sufficient enough.
+  * The semantics for retaining threads is a hint to the scheduler;
+    thus, no real guarantee is provided. The scheduler can ignore the hint and
+    move threads to another arena or to sleep if conditions are met.
+* End of a parallel phase:
+  * Indicates the point from which the scheduler may drop the hint and
+    no longer retain threads in the arena.
+  * Indicates that worker threads should avoid busy-waiting once there is no more work in the arena.
+    * Temporarily overrides the default arena leave policy, which will be restored when
+      new work is submitted.
+
+
+### Proposed API
+
+Summary of API changes:
+
+* Add enumeration class for the arena leave policy.
+* Add the policy as the last parameter to the arena constructor and initializer
+defaulted to "automatic".
+* Add functions to start and end the parallel phase to the `task_arena` class
+and the `this_task_arena` namespace.
+* Add RAII class to map a parallel phase to a code scope.
+
+```cpp
+class task_arena {
+    enum class leave_policy : /* unspecified type */ {
+        automatic = /* unspecifed */,
+        fast = /* unspecifed */,
+    };
+
+    task_arena(int max_concurrency = automatic, unsigned reserved_for_masters = 1,
+               priority a_priority = priority::normal,
+               leave_policy a_leave_policy = leave_policy::automatic);
+
+    task_arena(const constraints& constraints_, unsigned reserved_for_masters = 1,
+               priority a_priority = priority::normal,
+               leave_policy a_leave_policy = leave_policy::automatic);
+
+    void initialize(int max_concurrency, unsigned reserved_for_masters = 1,
+                    priority a_priority = priority::normal,
+                    leave_policy a_leave_policy = leave_policy::automatic);
+
+    void initialize(constraints a_constraints, unsigned reserved_for_masters = 1,
+                    priority a_priority = priority::normal,
+                    leave_policy a_leave_policy = leave_policy::automatic);
+
+    void start_parallel_phase();
+    void end_parallel_phase(bool with_fast_leave = false);
+
+    class scoped_parallel_phase {
+        scoped_parallel_phase(task_arena& ta, bool with_fast_leave = false);
+    };
```

Comment:
Looks like the scoped object can only be created for a known task_arena. How to use it with "this task arena"? Is it necessary to always "materialize" the arena with the help of the attaching constructor? Or should there be a special constructor for this class, either with no parameters or maybe with `tbb::attach` parameter?

Reply:
`tbb::attach` is [listed as deprecated interface](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onetbb/source/deprecated/task_arena_attach_tag) in the spec. Not sure why.

Reply:
@aleksei-fedotov I believe it is `tbb::task_arena::attach` being deprecated since there is now a [dedicated ](https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onetbb/source/task_scheduler/attach_tag_type)`tbb::attach` tag used for both `task_arena` and `task_scheduler_handle`.

Reply:
> Looks like the scoped object can only be created for a known task_arena. How to use it with "this task arena"? Is it necessary to always "materialize" the arena with the help of the attaching constructor? Or should there be a special constructor for this class, either with no parameters or maybe with a `tbb::attach` parameter?

With the current version of proposed API the only option is to create an arena with `tbb::attach` and then pass it to the `scoped_parallel_phase` constructor.
```cpp
tbb::task_arena ta{tbb::attach{}};
{
   tbb::task_arena::scoped_parallel_phase phase{ta};
   // parallel work
}
```
I guess it makes sense to improve usability and move this kind of logic under the hood. However, I am not sure about the way to approach it:
* I suppose using `tbb::attach` as a parameter makes sense when `scoped_parallel_phase` is within `task_arena` class scope.
* Having `tbb::attach` as a parameter when `scoped_parallel_phase` is within namespace `this_task_arena` scope seems excessive because `tbb::this_task_arena::scoped_parallel_phase` with no parameters is already self-explanatory.

Should we just introduce a new constructor with `tbb::attach` to `tbb::task_arena::scoped_parallel_phase` or should we have a separate `tbb::this_task_arena::scoped_parallel_phase` which doesn't accept `task_arena` at all?

Reply:
I prefer not to introduce an almost identical class. Currently, neither the `this_task_arena` namespace nor the similar `std::this_thread` namespace contain classes, but only functions.

So I prefer a special constructor, and I slightly prefer it to have the attach parameter, though a no-parameter constructor would probably be OK too.

Another question is: what this new constructor would do when an object is created in a thread that does not yet have any associated task arena?

Keeping these questions open and addressing later would be fine for me as well.

Reply:
Documented these questions in "Open Questions in Design".